nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,439 @@
+    sagactx: NexusActionContext,
+) -> Result<instance::UpdaterLock, ActionError> {
+    let RealParams {
+        ref serialized_authn, ref authz_instance, orig_lock, ..
+    } = sagactx.saga_params::<RealParams>()?;
+    let saga_id = sagactx.lookup::<Uuid>(INSTANCE_LOCK_ID)?;
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, serialized_authn);
+    let osagactx = sagactx.user_data();
+
+    slog::debug!(
+        osagactx.log(),
+        "instance update: trying to become instance updater...";
+        "instance_id" => %authz_instance.id(),
+        "saga_id" => %saga_id,
+        "parent_lock" => ?orig_lock,
+    );
+
+    let lock = osagactx
+        .datastore()
+        .instance_updater_inherit_lock(
+            &opctx,
+            &authz_instance,
+            orig_lock,
+            saga_id,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    slog::info!(
+        osagactx.log(),
+        "Now, I am become Updater, the destroyer of VMMs.";
+        "instance_id" => %authz_instance.id(),
+        "saga_id" => %saga_id,
+    );
+
+    Ok(lock)
+}
+
+async fn siu_unbecome_updater(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let RealParams { ref serialized_authn, ref authz_instance, .. } =
+        sagactx.saga_params::<RealParams>()?;
+    unlock_instance_inner(serialized_authn, authz_instance, &sagactx).await?;
+
+    Ok(())
+}
+
+async fn siu_migration_update_instance(
+    sagactx: NexusActionContext,
+) -> Result<PropolisUuid, ActionError> {
+    let RealParams { ref authz_instance, ref state, .. } =
+        sagactx.saga_params()?;
+    let migration = sagactx.lookup::<Migration>(MIGRATION)?;
+
+    let osagactx = sagactx.user_data();
+    let instance_id = InstanceUuid::from_untyped_uuid(authz_instance.id());
+
+    let mut new_runtime = state.instance.runtime().clone();
+    new_runtime.gen = Generation(new_runtime.gen.next());
+    new_runtime.time_updated = Utc::now();
+
+    // Determine how to update the instance record to reflect the current
+    // migration state.
+    let failed = migration.either_side_failed();
+    // If the migration has failed, or if the target reports that the migration
+    // has completed, clear the instance record's migration IDs so that a new
+    // migration can begin.
+    if failed || migration.target_state == MigrationState::COMPLETED {
+        info!(
+            osagactx.log(),
+            "instance update (migration {}): clearing migration IDs",
+            if failed { "failed" } else { "target_completed" };
+            "instance_id" => %instance_id,
+            "migration_id" => %migration.id,
+            "src_propolis_id" => %migration.source_propolis_id,
+            "target_propolis_id" => %migration.target_propolis_id,
+            "instance_update" => %"migration",
+        );
+        new_runtime.migration_id = None;
+        new_runtime.dst_propolis_id = None;
```

Comment:
The abandoned VMM background task is _supposed_ to handle this case, but I think that's something that we'll want to test.

We should probably have the `notify_instance_updated` function (which maybe ought to be renamed...) kick the abandoned VMM reaper task when it sees an unlinked VMM reach `Destroyed`/`Failed` so that it gets cleaned up in a timely manner. The instance-update saga _could_ also do that when it abandons a VMM, but that VMM may not actually _be_ in a deletable state yet when it's abandoned...